File: src/main/java/ru/umeta/libraryintegration/service/DocumentService.java
package ru.umeta.libraryintegration.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.util.StringUtils;
import ru.umeta.libraryintegration.dao.DocumentDao;
import ru.umeta.libraryintegration.dao.EnrichedDocumentDao;
import ru.umeta.libraryintegration.json.ModsParseResult;
import ru.umeta.libraryintegration.json.ParseResult;
import ru.umeta.libraryintegration.json.UploadResult;
import ru.umeta.libraryintegration.model.Document;
import ru.umeta.libraryintegration.model.EnrichedDocument;
import ru.umeta.libraryintegration.parser.ModsXMLParser;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.locks.ReentrantLock;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Created by ctash on 28.04.2015.
 */
public class DocumentService {

    private static final String DEFAULT_PROTOCOL = "Z39.50";
    private static final int DUPLICATE_SIZE = 1000;

    @Autowired
    private StringHashService stringHashService;

    @Autowired
    private ProtocolService protocolService;

    @Autowired
    private EnrichedDocumentDao enrichedDocumentDao;

    @Autowired
    private DocumentDao documentDao;

    @Autowired
    private ModsXMLParser modsXMLParser;

    private static class YearLocker {

        private static final String LOCKED = " Locked ";
        private static final String LOCKED_NULL = " Locked null";
        private final ConcurrentMap<Integer, ReentrantLock> lockMap = new ConcurrentHashMap<>();
        ReentrantLock lockForNull = new ReentrantLock();

        private void lockYear(Integer year) {

            if (year != null) {
                ReentrantLock lockForYear = new ReentrantLock();

                ReentrantLock mapLockForYear = lockMap.putIfAbsent(year, lockForYear);
                if (mapLockForYear != null) {
                    lockForYear = mapLockForYear;
                }

                lockForNull.lock();
                logThreadWithMessage(LOCKED_NULL);
                lockForYear.lock();
                logThreadWithMessage(LOCKED + year);
            } else {
                lockForNull.lock();
                logThreadWithMessage(LOCKED_NULL);
                lockMap.forEach((key, lock) -> {
                    lock.lock();
                    logThreadWithMessage(LOCKED + key);
                });
            }

        }

        private void logThreadWithMessage(String message) {
            //System.out.println(Thread.currentThread().getName() + message);
        }

        private void unlockYear(Integer year) {
            if (year != null) {
                lockMap.get(year).unlock();
                safeUnlock(lockForNull, null);
            } else {
                lockMap.forEach((key, lock) -> {
                    safeUnlock(lock, key);
                });
                safeUnlock(lockForNull, null);
            }
        }

        private void safeUnlock(ReentrantLock lock, Integer year) {
            if (lock.getHoldCount() > 0) {
                logThreadWithMessage(" Locked with hold count " + lock.getHoldCount());
                lock.unlock();
                if (year != null) {
                    logThreadWithMessage(" Unlocked " + year);
                } else {
                    logThreadWithMessage(" Unlocked null");
                }
            } else {
                logThreadWithMessage(" Tried to unlock with 0 hold count. Did nothing " + year);
            }
        }

        private void unlockOnlyNull() {
            safeUnlock(lockForNull, null);
        }

    }

    private final YearLocker locker = new YearLocker();

    public UploadResult processDocumentList(List<ParseResult> resultList, String protocolName) {
        int newEnriched = 0;
        int parsedDocs = 0;
        List<Callable<Object>> threadList = new ArrayList<>();
        ExecutorService executor = Executors.newFixedThreadPool(10);
        for (ParseResult parseResult : checkNotNull(resultList)) {
            if (parseResult instanceof ModsParseResult) {
                try {
                    Document document = new Document();
                    ModsParseResult modsParseResult = (ModsParseResult) parseResult;

                    Thread thread = new Thread(() -> {
                        try {
                            Integer publishYear = modsParseResult.getPublishYear();
                            if (publishYear != null && publishYear.equals(-1)) {
                                publishYear = null;
                            }
                            locker.lockYear(publishYear);

                            document.setAuthor(stringHashService.getFromRepository(modsParseResult.getAuthor()));
                            document.setTitle(stringHashService.getFromRepository(modsParseResult.getTitle()));
                            document.setCreationTime(new Date());
                            String isbn = modsParseResult.getIsbn();
                            if (StringUtils.isEmpty(isbn)) {
                                isbn = null;
                            }
                            document.setIsbn(isbn);
                            document.setProtocol(protocolService.getFromRepository(protocolName == null ? DEFAULT_PROTOCOL : protocolName));
                            document.setPublishYear(publishYear);
                            try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
                                modsParseResult.getModsDefinition().save(outputStream);
                                document.setXml(new String(outputStream.toByteArray(), "UTF-8"));
                            } catch (IOException e) {
                                e.printStackTrace();
                                return;
                            }
                            EnrichedDocument enrichedDocument = findEnrichedDocument(document);
                            if (enrichedDocument != null) {
                                if (enrichedDocument.getPublishYear() == null) {
                                    enrichedDocument.setPublishYear(document.getPublishYear());
                                }
                                if (enrichedDocument.getIsbn() == null) {
                                    enrichedDocument.setIsbn(document.getIsbn());
                                }
                                mergeDocuments(modsParseResult, enrichedDocument);
                                enrichedDocumentDao.saveOrUpdate(enrichedDocument);
                            } else {
                                enrichedDocument = new EnrichedDocument();
                                enrichedDocument.setAuthor(document.getAuthor());
                                enrichedDocument.setTitle(document.getTitle());
                                enrichedDocument.setIsbn(document.getIsbn());
                                enrichedDocument.setXml(document.getXml());
                                enrichedDocument.setCreationTime(document.getCreationTime());
                                enrichedDocument.setPublishYear(document.getPublishYear());
                                enrichedDocument.setId(enrichedDocumentDao.save(enrichedDocument).longValue());
                                //newEnriched++;
                                document.setDistance(1.);
                            }
                            document.setEnrichedDocument(enrichedDocument);
                            documentDao.save(document);

                            //System.out.println(Thread.currentThread().getName() + " Unlocking publish year " + publishYear);
                            locker.unlockYear(publishYear);
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                        //parsedDocs++;
                    });

                    threadList.add(Executors.callable(thread));
                } catch (Exception e) {
                    System.err.println("ERROR. Failed to add a document with title {" +
                            parseResult.getTitle() + "}, author {" +
                            parseResult.getAuthor() + "}");
                }


            }
        }
        try {
            executor.invokeAll(threadList);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return new UploadResult(parsedDocs, newEnriched);
    }

    private void mergeDocuments(ModsParseResult modsParseResult, EnrichedDocument enrichedDocument) {
        modsXMLParser.enrich(modsParseResult.getModsDefinition(), enrichedDocument);
    }

    private EnrichedDocument findEnrichedDocument(Document document) {

        //first check whether the document has isbn or not
        String isbn = document.getIsbn();
        Integer publishYear = document.getPublishYear();
        List<EnrichedDocument> nearDuplicates;
        if (isbn == null && publishYear == null) {
            // if it's null, we search through every record in the storage
            nearDuplicates = enrichedDocumentDao.getNearDuplicates(document);
        } else {

            // if it's not null, we search through record where isbn is the same
            if (publishYear == null) {
                nearDuplicates = enrichedDocumentDao.getNearDuplicatesWithIsbn(document);

                if (nearDuplicates == null || nearDuplicates.size() == 0) {
                    // if it didn't find anything, search through record with null isbn.
                    nearDuplicates = enrichedDocumentDao.getNearDuplicatesWithNullIsbn(document);
                }
            } else if (isbn == null) {
                nearDuplicates = enrichedDocumentDao.getNearDuplicatesWithPublishYear(document);

            } else {
                //both publishYear and isbn is not null
                nearDuplicates = enrichedDocumentDao.getNearDuplicatesWithIsbnAndPublishYear(document);

                if (nearDuplicates == null || nearDuplicates.size() == 0) {
                    // if it didn't find anything, search through record with null isbn.
                    nearDuplicates = enrichedDocumentDao.getNearDuplicatesWithIsbn(document);

                    if (nearDuplicates == null || nearDuplicates.size() == 0) {
                        nearDuplicates = enrichedDocumentDao.getNearDuplicatesWithPublishYear(document);
                    }
                }
            }

        }

        if (nearDuplicates != null && nearDuplicates.size() > 0) {

            double maxDistance = 0;
            double minDistance = 0.4;
            EnrichedDocument closestDocument = null;
            String titleValue = document.getTitle().getValue();
            String authorValue = document.getAuthor().getValue();

            //If the new document has a publish year and nearDuplicates does not contain null publish years
            //it means that we can unlock the null lock, as the data to be finally stored will have concrete year
            if (publishYear != null) {
                Optional<EnrichedDocument> first = nearDuplicates.stream().filter(
                        nearDuplicate -> nearDuplicate.getPublishYear() == null)
                        .findFirst();
                if (!first.isPresent()) {
                    //System.out.println(Thread.currentThread().getName() + " Unlocking only null publish year " + publishYear);
                    locker.unlockOnlyNull();
                }
            }

            for (EnrichedDocument nearDuplicate : nearDuplicates) {

                double titleDistance = stringHashService.distance(titleValue, nearDuplicate.getTitle().getValue());

                double authorDistance = stringHashService.distance(authorValue, nearDuplicate.getAuthor().getValue());

                double resultDistance = (titleDistance + authorDistance) / 2;

                if (resultDistance > maxDistance && resultDistance > minDistance) {
                    maxDistance = resultDistance;
                    closestDocument = nearDuplicate;
                }
            }

            if (closestDocument != null && closestDocument.getPublishYear() != null) {
                //System.out.println(Thread.currentThread().getName() + " Unlocking only null publish year " + publishYear);
                locker.unlockOnlyNull();
            }

            document.setDistance(maxDistance);
            return closestDocument;
        }

        return null;
    }

    public List<ParseResult> addNoise(ParseResult parseResult, int saltLevel) {
        String author = parseResult.getAuthor();
        String title = parseResult.getTitle();

        int authorLength = author.length();
        int titleLength = title.length();

        parseResult.setIsbn(null);
        if (StringUtils.isEmpty(author) || StringUtils.isEmpty(title)) {
            return null;
        } else {
            ArrayList<ParseResult> resultList = new ArrayList<>();
            for (int i = 0; i < DUPLICATE_SIZE; i++) {
                ParseResult newParseResult = parseResult.clone();
                StringBuilder newAuthor = new StringBuilder(author);
                StringBuilder newTitle = new StringBuilder(title);
                for (int j = 0; j < saltLevel; j++) {
                    Random rnd = new Random();
                    int saltIndex = rnd.nextInt(authorLength);
                    newAuthor.setCharAt(saltIndex, '#');

                    saltIndex = new Random().nextInt(titleLength);
                    newTitle.setCharAt(saltIndex, '#');
                }
                newParseResult.setAuthor(newAuthor.toString());
                newParseResult.setTitle(newTitle.toString());
                resultList.add(newParseResult);
            }
            return resultList;
        }
    }
}

File: src/main/java/ru/umeta/libraryintegration/service/DocumentService.kt
package ru.umeta.libraryintegration.service

import gnu.trove.set.hash.TLongHashSet
import org.springframework.util.StringUtils
import ru.umeta.libraryintegration.inmemory.EnrichedDocumentRepository
import ru.umeta.libraryintegration.json.ModsParseResult
import ru.umeta.libraryintegration.json.ParseResult
import ru.umeta.libraryintegration.json.UploadResult
import ru.umeta.libraryintegration.model.EnrichedDocument
import ru.umeta.libraryintegration.model.EnrichedDocumentLite
import java.util.*


/**
 * Service for operating with [Document] and [EnrichedDocument]
 * Created by ctash on 28.04.2015.
 */
object DocumentService : AutoCloseable {

    private val DEFAULT_PROTOCOL = "Z39.50"
    private val DUPLICATE_SIZE = 1000

    val enrichedDocumentRepository = EnrichedDocumentRepository
    val stringHashService = StringHashService

    fun processDocumentList(resultList: List<ParseResult>, protocolName: String?): UploadResult {
        var newEnriched = 0
        var parsedDocs = 0

        for (parseResult in resultList) {
            if (parseResult is ModsParseResult) {
                try {
                    var author = parseResult.author
                    if (author.length > 255) {
                        author = author.substring(0, 255)
                    }
                    var title = parseResult.title
                    if (title.length > 255) {
                        title = title.substring(0, 255)
                    }

                    val authorId = stringHashService.getFromRepository(author)
                    val titleId = stringHashService.getFromRepository(title)
                    var isbn: String? = parseResult.isbn
                    if (isbn.isNullOrEmpty()) {
                    }
                    isbn = null
                    val enrichedDocument = EnrichedDocument(-1, authorId, titleId, isbn, null, Date(),
                            parseResult.publishYear)
                    enrichedDocumentRepository.save(enrichedDocument)
                    newEnriched++;
                    parsedDocs++
                } catch (e: Exception) {
                    throw e
                }


            }
        }
        return UploadResult(parsedDocs, newEnriched)
    }

    fun findEnrichedDocuments(cur: EnrichedDocumentLite, used: TLongHashSet): DFS {
        val authorId = cur.authorId
        val titleId = cur.titleId
        val authorTokens = stringHashService.getById(authorId).tokens
        val titleTokens = stringHashService.getById(titleId).tokens
        used.add(cur.id)
        //filter documents which have the nearest measure of 0.7 or more
        var nearDuplicates = enrichedDocumentRepository.getNearDuplicates(cur, used)
        var iterationsLength = 0L
        var iterationsSetInter = 0L
        nearDuplicates = nearDuplicates.filter {
            if (used.contains(it.id)) {
                false
            } else {
                val itAuthorTokens = stringHashService.getById(it.authorId).tokens
                var authorTokensRatio: Double = authorTokens.size() * 1.0 / itAuthorTokens.size()
                if (authorTokensRatio > 1) {
                    authorTokensRatio = 1.0 / authorTokensRatio
                }

                if (authorTokensRatio < 0.4) {
                    iterationsLength++
                    false
                } else {
                    val itTitleTokens = stringHashService.getById(it.titleId).tokens
                    var titleTokensRatio: Double = titleTokens.size() * 1.0 / itTitleTokens.size()
                    if (titleTokensRatio > 1) {
                        titleTokensRatio = 1.0 / titleTokensRatio
                    }

                    if (authorTokensRatio + titleTokensRatio < 0.7 * 2) {
                        iterationsLength++
                        false
                    } else {
                        iterationsSetInter++
                        (stringHashService.distance(authorTokens, itAuthorTokens) + stringHashService.distance
                        (titleTokens, itTitleTokens) >= 0.7 * 2)
                    }
                }
            }
        }

        val dfs = DFS(nearDuplicates, iterationsLength, iterationsSetInter)
        return dfs
    }

    data class DFS(
        val component: List<EnrichedDocumentLite>,
        var iterationsLength: Long,
        var iterationsSetInter: Long)



    fun addNoise(parseResult: ParseResult, saltLevel: Int): List<ParseResult>? {
        val author = parseResult.author
        val title = parseResult.title

        val authorLength = author.length
        val titleLength = title.length

        parseResult.isbn = null
        if (StringUtils.isEmpty(author) || StringUtils.isEmpty(title)) {
            return null
        } else {
            val resultList = ArrayList<ParseResult>()
            for (i in 0..DUPLICATE_SIZE - 1) {
                val newParseResult = parseResult.clone()
                val newAuthor = StringBuilder(author)
                val newTitle = StringBuilder(title)
                for (j in 0..saltLevel - 1) {
                    val rnd = Random()
                    var noiseIndex = rnd.nextInt(authorLength)
                    newAuthor.setCharAt(noiseIndex, '#')

                    noiseIndex = Random().nextInt(titleLength)
                    newTitle.setCharAt(noiseIndex, '#')
                }
                newParseResult.author = newAuthor.toString()
                newParseResult.title = newTitle.toString()
                resultList.add(newParseResult)
            }
            return resultList
        }
    }

    override fun close() {
        enrichedDocumentRepository.close()
        stringHashService.close()
    }

    fun getDocuments(): List<EnrichedDocumentLite> {
        return enrichedDocumentRepository.getList();
    }

    fun processDocumentListInit(resultList: List<ParseResult>, nothing: Nothing?): Any {
        var newEnriched = 0
        var parsedDocs = 0

        for (parseResult in resultList) {
            if (parseResult is ModsParseResult) {
                try {
                    var author = parseResult.author
                    if (author.length > 255) {
                        author = author.substring(0, 255)
                    }
                    var title = parseResult.title
                    if (title.length > 255) {
                        title = title.substring(0, 255)
                    }

                    val authorId = stringHashService.getFromRepositoryInit(author)
                    val titleId = stringHashService.getFromRepositoryInit(title)
                    var isbn: String? = parseResult.isbn
                    if (isbn.isNullOrEmpty()) {
                    }
                    isbn = null
                    val enrichedDocument = EnrichedDocument(-1, authorId, titleId, isbn, null, Date(),
                            parseResult.publishYear)
                    enrichedDocumentRepository.saveInit(enrichedDocument)
                    newEnriched++;
                    parsedDocs++
                } catch (e: Exception) {
                    throw e
                }


            }
        }
        return UploadResult(parsedDocs, newEnriched)
    }
}
